### ISSUE TITLE
Extract App::respond() to a separate class

### ISSUE THREAD

**Issue by stivni** [slimphp/Slim#1935]
Extract App::respond() to a class that can handle the responding.
It is conform single responsibility principle, would make the App class less long and make the class usable on its own.

**Comment by geggleto** [comment]
The app's job is to process the response and respond to the browser.

While I agree that the App class is long and we can improve it... responding to a request is the primary job of the app class.

Also any change made would need to be for 4.x ... where I suspect we can make significant gains in the overall flow of the class.

**Comment by stivni** [comment]
> The app's job is to process the response and respond to the browser.

In ideal circumstances, I would indeed only use the App, but in legacy systems, I probably want to start using parts separately until all pieces of the puzzle fall in place.

> responding to a request is the primary job of the app class.

I'm not asking to isolate how the app responds to a request, but how the response itself is put on the wire. In a true object oriented way, we would ask the response to output itself or at least output itself to a given output stream. Because PSR-7 we don't have any method for that to implement, so it would make sense to have an isolated part of the application that does that.

And just something philosophical: in the end all jobs will be performed within an app, so does that mean that all code should be in the same class?

**Comment by geggleto** [comment]
> And just something philosophical: in the end all jobs will be performed within an app, so does that mean that all code should be in the same class?

I guess we could provide a Psr7-Responder as a community project... It wasn't abstracted out because there's no reason to. We wouldn't want to swap out implementations.

**Comment by geggleto** [comment]
There's no harm in it being in a different class... there's just no added benefit aside from marginally easier  unit testing for code coverage.

**Comment by akrabat** [comment]
The `Responder` class should be part of [Slim-Http](https://github.com/slimphp/Slim-Http).

It should contain the code from `App::finalize()` and `App::respond()`. `App::handle()` would then use the Responder something like this:

```php
    public function handle(ServerRequestInterface $request): ResponseInterface
    {
        //... 

        $response = $this->callMiddlewareStack($request, $response);
        
        $responder = $this->getResponder();
        return $responder->respond($response);
    }
```

**Comment by potievdev** [comment]
I sent PR https://github.com/slimphp/Slim-Http/pull/72

**Comment by l0gicgate** [comment]
This issue will be resolved when merging #2529
